scikit-build-core breaks building C extensions via pdm-backend's setuptools use

pdm-backend can optionally use setuptools to build C extensions. However, when scikit-build-core is installed, the setuptools becomes broken somehow and C extensions aren't built (this is a different bug than #413, and I can reproduce with 3e1f31782be7f80921b707a171d3ddb830567c41).

To reproduce:

```
$ git clone https://github.com/pdm-project/pdm-backend/
$ pip install . editables pytest setuptools
[…]
$ python -m pytest
[…]
=================================================== 63 passed, 9 warnings in 2.05s ====================================================
$ pip install scikit-build-core
[…]
$ python -m pytest
========================================================= test session starts =========================================================
platform linux -- Python 3.12.0b3, pytest-7.4.0, pluggy-1.2.0
rootdir: /tmp/pdm-backend
collected 63 items                                                                                                                    

tests/test_api.py .........FF..................                                                                                 [ 46%]
tests/test_file_finder.py ....................                                                                                  [ 77%]
tests/test_hooks.py .                                                                                                           [ 79%]
tests/test_metadata.py .........                                                                                                [ 93%]
tests/test_utils.py .                                                                                                           [ 95%]
tests/test_wheel.py ...                                                                                                         [100%]

============================================================== FAILURES ===============================================================
_____________________________________________________ test_build_with_cextension ______________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-mgorny/pytest-7/test_build_with_cextension0')

    def test_build_with_cextension(tmp_path: Path) -> None:
        with build_fixture_project("demo-cextension"):
            wheel_name = api.build_wheel(tmp_path.as_posix())
            sdist_name = api.build_sdist(tmp_path.as_posix())
            assert api.get_requires_for_build_sdist() == []
            assert api.get_requires_for_build_wheel() == ["setuptools>=40.8.0"]
    
            zip_names = get_wheel_names(tmp_path / wheel_name)
            assert "my_package/__init__.py" in zip_names
            assert (
                "my_package/hellomodule.c" not in zip_names
            ), "Not collect c files while building wheel"
            extension_suffix = ".pyd" if sys.platform == "win32" else ".so"
>           assert any(name.endswith(extension_suffix) for name in zip_names)
E           assert False
E            +  where False = any(<generator object test_build_with_cextension.<locals>.<genexpr> at 0x7fa12d1635a0>)

tests/test_api.py:185: AssertionError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
running build
running build_py
creating build
creating build/lib.linux-x86_64-cpython-312
creating build/lib.linux-x86_64-cpython-312/my_package
copying my_package/__init__.py -> build/lib.linux-x86_64-cpython-312/my_package
running egg_info
writing demo_package.egg-info/PKG-INFO
writing dependency_links to demo_package.egg-info/dependency_links.txt
writing top-level names to demo_package.egg-info/top_level.txt
reading manifest file 'demo_package.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'demo_package.egg-info/SOURCES.txt'
copying my_package/hellomodule.c -> build/lib.linux-x86_64-cpython-312/my_package
running build_ext
building 'my_package.hello' extension
creating build/temp.linux-x86_64-cpython-312
creating build/temp.linux-x86_64-cpython-312/my_package
x86_64-pc-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -fPIC -I/tmp/pdm-backend/venv/include -I/usr/include/python3.12 -c my_package/hellomodule.c -o build/temp.linux-x86_64-cpython-312/my_package/hellomodule.o
x86_64-pc-linux-gnu-gcc -shared build/temp.linux-x86_64-cpython-312/my_package/hellomodule.o -L/usr/lib64 -o build/lib.linux-x86_64-cpython-312/my_package/hello.cpython-312-x86_64-linux-gnu.so
 - Adding .pdm-build/.gitignore -> .gitignore
 - Adding .pdm-build/demo_package-0.1.0.dist-info/METADATA -> demo_package-0.1.0.dist-info/METADATA
 - Adding .pdm-build/demo_package-0.1.0.dist-info/WHEEL -> demo_package-0.1.0.dist-info/WHEEL
 - Adding .pdm-build/demo_package-0.1.0.dist-info/licenses/LICENSE -> demo_package-0.1.0.dist-info/licenses/LICENSE
 - Adding my_package/__init__.py -> my_package/__init__.py
 - Adding demo_package-0.1.0.dist-info/RECORD -> demo_package-0.1.0.dist-info/RECORD
 - Adding LICENSE -> LICENSE
 - Adding build/lib.linux-x86_64-cpython-312/my_package/hellomodule.c -> build/lib.linux-x86_64-cpython-312/my_package/hellomodule.c
 - Adding my_package/__init__.py -> my_package/__init__.py
 - Adding my_package/hellomodule.c -> my_package/hellomodule.c
 - Adding pdm_build.py -> pdm_build.py
 - Adding .pdm-build/pyproject.toml -> pyproject.toml
 - Adding PKG-INFO -> PKG-INFO
__________________________________________________ test_build_with_cextension_in_src __________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-mgorny/pytest-7/test_build_with_cextension_in_0')

    def test_build_with_cextension_in_src(tmp_path: Path) -> None:
        with build_fixture_project("demo-cextension-in-src"):
            wheel_name = api.build_wheel(tmp_path.as_posix())
            sdist_name = api.build_sdist(tmp_path.as_posix())
    
            zip_names = get_wheel_names(tmp_path / wheel_name)
            assert "my_package/__init__.py" in zip_names
            assert (
                "my_package/hellomodule.c" not in zip_names
            ), "Not collect c files while building wheel"
            extension_suffix = ".pyd" if sys.platform == "win32" else ".so"
>           assert any(name.endswith(extension_suffix) for name in zip_names)
E           assert False
E            +  where False = any(<generator object test_build_with_cextension_in_src.<locals>.<genexpr> at 0x7fa12d2382e0>)

tests/test_api.py:208: AssertionError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
running build
running build_py
creating build
creating build/lib.linux-x86_64-cpython-312
creating build/lib.linux-x86_64-cpython-312/my_package
copying src/my_package/__init__.py -> build/lib.linux-x86_64-cpython-312/my_package
running egg_info
writing src/demo_package.egg-info/PKG-INFO
writing dependency_links to src/demo_package.egg-info/dependency_links.txt
writing top-level names to src/demo_package.egg-info/top_level.txt
reading manifest file 'src/demo_package.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'src/demo_package.egg-info/SOURCES.txt'
copying src/my_package/hellomodule.c -> build/lib.linux-x86_64-cpython-312/my_package
running build_ext
building 'my_package.hello' extension
creating build/temp.linux-x86_64-cpython-312
creating build/temp.linux-x86_64-cpython-312/src
creating build/temp.linux-x86_64-cpython-312/src/my_package
x86_64-pc-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -fPIC -I/tmp/pdm-backend/venv/include -I/usr/include/python3.12 -c src/my_package/hellomodule.c -o build/temp.linux-x86_64-cpython-312/src/my_package/hellomodule.o
x86_64-pc-linux-gnu-gcc -shared build/temp.linux-x86_64-cpython-312/src/my_package/hellomodule.o -L/usr/lib64 -o build/lib.linux-x86_64-cpython-312/my_package/hello.cpython-312-x86_64-linux-gnu.so
 - Adding .pdm-build/.gitignore -> .gitignore
 - Adding .pdm-build/demo_package-0.1.0.dist-info/METADATA -> demo_package-0.1.0.dist-info/METADATA
 - Adding .pdm-build/demo_package-0.1.0.dist-info/WHEEL -> demo_package-0.1.0.dist-info/WHEEL
 - Adding .pdm-build/demo_package-0.1.0.dist-info/licenses/LICENSE -> demo_package-0.1.0.dist-info/licenses/LICENSE
 - Adding src/my_package/__init__.py -> my_package/__init__.py
 - Adding demo_package-0.1.0.dist-info/RECORD -> demo_package-0.1.0.dist-info/RECORD
 - Adding LICENSE -> LICENSE
 - Adding build/lib.linux-x86_64-cpython-312/my_package/hellomodule.c -> build/lib.linux-x86_64-cpython-312/my_package/hellomodule.c
 - Adding pdm_build.py -> pdm_build.py
 - Adding .pdm-build/pyproject.toml -> pyproject.toml
 - Adding src/my_package/__init__.py -> src/my_package/__init__.py
 - Adding src/my_package/hellomodule.c -> src/my_package/hellomodule.c
 - Adding PKG-INFO -> PKG-INFO
========================================================== warnings summary ===========================================================
tests/test_api.py::test_build_wheel_write_version_to_file
tests/test_api.py::test_build_wheel_write_version_to_file_template
tests/test_api.py::test_clean_not_called_if_envset[true-False]
tests/test_api.py::test_clean_not_called_if_envset[false-True]
tests/test_api.py::test_clean_not_called_if_envset[0-True]
tests/test_api.py::test_clean_not_called_if_envset[1-False]
tests/test_api.py::test_clean_not_called_if_config_settings_exist[-False]
tests/test_api.py::test_clean_not_called_if_config_settings_exist[True-False]
tests/test_api.py::test_clean_not_called_if_config_settings_exist[None-False]
  /tmp/pdm-backend/venv/lib/python3.12/site-packages/pdm/backend/hooks/version/scm.py:326: DeprecationWarning: datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.now(datetime.UTC).
    local_version = fmt.format(node=version.node, time=datetime.utcnow())

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= short test summary info =======================================================
FAILED tests/test_api.py::test_build_with_cextension - assert False
FAILED tests/test_api.py::test_build_with_cextension_in_src - assert False
============================================== 2 failed, 61 passed, 9 warnings in 2.16s ===============================================
```

CC @frostming

--- a/src/scikit_build_core/setuptools/build_cmake.py
+++ b/src/scikit_build_core/setuptools/build_cmake.py
@@ -197,7 +197,7 @@ def _prepare_extension_detection(dist: Distribution) -> None:
 
 def _prepare_build_cmake_command(dist: Distribution) -> None:
     # Prepare new build_cmake command and make sure build calls it
-    build = dist.get_command_obj("build")
+    build = dist.get_command_class("build")
     assert build is not None
     if "build_cmake" not in {x for x, _ in build.sub_commands}:
         build.sub_commands.append(

